numba/tests/npyufunc/test_dufunc.py

```diff
@@ -85,6 +96,125 @@ def test_ufunc_props(self):
+
+    def test_ufunc_props_jit(self):
+        duadd = self.nopython_dufunc(pyuadd)
+        duadd(1, 2)  # initialize types attribute
+
+        attributes = {'nin': duadd.nin,
+                      'nout': duadd.nout,
+                      'nargs': duadd.nargs,
+                      #'ntypes': duadd.ntypes,
+                      #'types': duadd.types,
+                      'identity': duadd.identity,
+                      'signature': duadd.signature}
+
+        def get_attr_fn(attr):
+            fn = f'''
+                def impl():
+                    return duadd.{attr}
+            '''
+            l = {}
+            exec(textwrap.dedent(fn), {'duadd': duadd}, l)
+            return l['impl']
+
+        for attr, val in attributes.items():
+            cfunc = njit(get_attr_fn(attr))
+            self.assertEqual(val, cfunc(), f'attribute: {attr}')
+
+        # We don't expose [n]types attributes as they are dynamic attributes
+        # and can change as the user calls the ufunc
+        # cfunc = njit(get_attr_fn('ntypes'))
+        # self.assertEqual(cfunc(), 1)
+        # duadd(1.1, 2.2)
+        # self.assertEqual(cfunc(), 2)
+
+
+class TestDUFuncMethods(TestCase):
+    def _check_reduce(self, ufunc, dtype=None, initial=None):
+
+        @njit
+        def foo(a, axis, dtype, initial):
+            return ufunc.reduce(a,
+                                axis=axis,
+                                dtype=dtype,
+                                initial=initial)
+
+        inputs = [
+            np.arange(5),
+            np.arange(4).reshape(2, 2),
+            np.arange(40).reshape(5, 4, 2),
+        ]
+        for array in inputs:
+            for axis in range(array.ndim):
+                expected = foo.py_func(array, axis, dtype, initial)
+                got = foo(array, axis, dtype, initial)
+                self.assertPreciseEqual(expected, got)
+
+    def _check_reduce_axis(self, ufunc, dtype, initial=None):
+
+        @njit
+        def foo(a, axis):
+            return ufunc.reduce(a, axis=axis, initial=initial)
+
+        exc_msg = (f"reduction operation '{ufunc.__name__}' is not "
+                   "reorderable, so at most one axis may be specified")
+        inputs = [
+            np.arange(40, dtype=dtype).reshape(5, 4, 2),
+            np.arange(10, dtype=dtype),
+        ]
+        for array in inputs:
+            for i in range(1, array.ndim + 1):
+                for axis in itertools.combinations(range(array.ndim), r=i):
+                    try:
+                        expected = foo.py_func(array, axis)
+                    except ValueError as e:
+                        self.assertEqual(e.args[0], exc_msg)
+                        with self.assertRaisesRegex(TypingError, exc_msg):
+                            got = foo(array, axis)
+                    else:
+                        got = foo(array, axis)
+                        self.assertPreciseEqual(expected, got)
```

Comment:
Not sure if this is the correct way for error testing. We can just check error throwing once in a different function, the rest of the cases here can be simply all valid cases.

Reply:
Problem is that some of the combinations generated by `itertools.combinations` are invalid, and it is not possible to know that beforehand